File: src/commands/item.rs
use crate::models::Slot;
use crate::utils::{build_item_editor_response, find_item_by_name};
use crate::{Context, Error, EMPTY_ENCHANTMENT_ID, EMPTY_MODIFIER_ID};

/// Get information about an item with interactive editor
#[poise::command(slash_command)]
pub async fn item(
    ctx: Context<'_>,
    #[description = "Name of the item"]
    #[autocomplete = "autocomplete_item"]
    name: String,
) -> Result<(), Error> {
    let data = ctx.data();
    let item = match find_item_by_name(data, &name) {
        Some(item) => item,
        None => {
            let response = poise::CreateReply::default()
                .content("❌ Item not found!")
                .ephemeral(true);
            ctx.send(response).await?;
            return Ok(());
        }
    };

    // Create initial slot with max level
    let slot = Slot {
        item: item.id.clone(),
        enchant: EMPTY_ENCHANTMENT_ID.to_string(),
        modifier: EMPTY_MODIFIER_ID.to_string(),
        gems: vec![],
        level: crate::MAX_LEVEL,
    };

    let (embed, components) = build_item_editor_response(&slot, data).await;

    let response = poise::CreateReply::default()
        .embed(embed)
        .components(components);

    ctx.send(response).await?;
    Ok(())
}

async fn autocomplete_item(ctx: Context<'_>, partial: &str) -> impl Iterator<Item = String> {
    let data = ctx.data();
    let items = data.items_data.read();

    items
        .iter()
        .filter(|item| {
            item.name.to_lowercase().contains(&partial.to_lowercase()) && item.name != "None"
        })
        .take(25)
        .map(|item| item.name.clone())
        .collect::<Vec<_>>()
        .into_iter()
}

File: src/commands/item.ts
import {
  ActionRow,
  Button,
  Command,
  CommandContext,
  createStringOption,
  Declare,
  Embed,
  Options,
} from "seyfert";
import { getData } from "../data/load";
import type { TotalStats } from "../types";
import { EMBED_FOOTER } from "../constants";
import { calculateItemStats, formatTotalStats } from "../utils";
import { getRarityColor } from "../utils";
import { ButtonStyle } from "seyfert/lib/types";

const options = {
  name: createStringOption({
    description: "The name of the item to search for",
    required: true,
    autocomplete: async (interaction) => {
      const itemsData = (await getData()).items;
      const focus = interaction.getInput();
      const response = Object.values(itemsData)
        .filter(
          (val) =>
            val.name.toLowerCase().includes(focus.toLowerCase()) &&
            val.name !== "None"
        )
        .slice(0, 25)
        .map((val) => ({ name: val.name, value: val.id }));

      return interaction.respond(response);
    },
  }),
};

@Declare({
  name: "item",
  description: "Gets the stats of an item given its name.",
})
@Options(options)
export default class ItemCommand extends Command {
  override async run(ctx: CommandContext<typeof options>) {
    const response = ctx.options.name;
    const itemsData = (await getData()).items;
    const item = itemsData[response];

    if (!item) {
      await ctx.write({
        content: `Error: selected item does not exist: ${response}`,
      });
      return;
    }

    const totalStats: TotalStats = calculateItemStats(item);

    const formattedStats = formatTotalStats(totalStats);

    const embed = new Embed()
      .setAuthor({
        name: ctx.author.username,
        iconUrl: ctx.author.avatarURL(),
      })
      .setThumbnail(item.imageId)
      .setTitle(`${item.name} | ${item.id}`)
      .setColor(getRarityColor(item.rarity))
      .setFooter({ text: EMBED_FOOTER })
      .setFields([
        {
          name: "Description",
          value: item.legend,
        },
        {
          name: "Stats",
          value: formattedStats,
        },
        {
          name: "Type",
          value: item.mainType,
          inline: true,
        },
        {
          name: "Subtype",
          value: item.subType ?? "None",
          inline: true,
        },
        {
          name: "Rarity",
          value: item.rarity,
          inline: true,
        },
      ]);

    const components = [
      new Button()
        .setLabel("Add Enchant")
        .setCustomId("item_set_enchant")
        .setStyle(ButtonStyle.Secondary),
    ];

    if (item.validModifiers) {
      components.push(
        new Button()
          .setLabel("Add Modifier")
          .setCustomId("item_set_modifier")
          .setStyle(ButtonStyle.Secondary)
      );
    }

    if (item.gemNo) {
      components.push(
        new Button()
          .setLabel("Add Gems")
          .setCustomId("item_set_gems")
          .setStyle(ButtonStyle.Secondary)
      );
    }

    const row = new ActionRow().setComponents(components);

    await ctx.write({
      embeds: [embed],
      components: [row],
    });
  }
}
